Raise on non-existent kwargs

smcabrera opened the pull request:
closes https://github.com/AaronLasseigne/active_interaction/issues/585

smcabrera commented on lib/active_interaction/base.rb:
```diff
@@ -36,6 +36,20 @@ class << self
       include Hashable
       include Missable
 
+      ALLOWED_KWARGS = %i[
+        desc
+        default
+        index_errors
+        strip
+        format
+        digits
+        base
+        from
+        methods
+        class
+        converter
+      ].freeze
```
These were all the options I could find from the those mentioned in the docs.

Not all of these options are possible for every kind of filter; a more sophisticated version of this would verify that the given option was possible for the filter, but I wanted to stand up a PR to see what people think first.

smcabrera commented on lib/active_interaction/base.rb:
```diff
@@ -105,6 +119,9 @@ def method_missing(*args, &block)
       def add_filter(klass, name, options, &block)
         raise InvalidFilterError, %("#{name}" is a reserved name) if Inputs.reserved?(name)
+        if (invalid_options = options.keys - ALLOWED_KWARGS).any?
+          raise InvalidFilterError, "invalid options: #{invalid_options.join(', ')}"
```
Is the right exception to use? Also I wanted to make sure the options were in the message to aid in debugging but not sure if the format should change.

antulik replied:
Filter option validation is more suitable for Filter classes as each filter can validate it's own options. Consider moving them there.

smcabrera commented on lib/active_interaction/filters/hash_filter.rb:
```diff
@@ -77,7 +77,7 @@ def convert(value)
     # rubocop:disable Style/MissingRespondToMissing
     def method_missing(*args, &block)
-      super(*args) do |klass, names, options|
+      super do |klass, names, options|
```
I ran `rake` and rubocop complained about this one, although it seems unrelated to any of my changes, not sure why it wouldn't have come up before 🤔

AaronLasseigne replied:
I can look into this.

smcabrera commented on spec/active_interaction/errors_spec.rb:
```diff
@@ -11,8 +11,8 @@
     Class.new(ActiveInteraction::Base) do
-      string :attribute, defualt: nil
-      array :array, defualt: nil
+      string :attribute, default: nil
+      array :array, default: nil
```
I wasn't sure if this misspelling was intentional since the spec is about errors. But it still passes after fixing it (and my change would break it). @AaronLasseigne you would know better than I would.

AaronLasseigne replied:
Not intentional 😬, maybe this PR is a good idea!

antulik commented on lib/active_interaction/base.rb:
```diff
@@ -105,6 +119,9 @@ def method_missing(*args, &block)
       # @param options [Hash]
       def add_filter(klass, name, options, &block)
         raise InvalidFilterError, %("#{name}" is a reserved name) if Inputs.reserved?(name)
+        if (invalid_options = options.keys - ALLOWED_KWARGS).any?
```
Comment on the implementation specifically. Using constant ALLOWED_KWARGS makes it impossible to extend functionality with new behavior. Consider using methods, as they can be extended.

AaronLasseigne commented:
Sorry I've been slow to get to this. Life and all. I like the idea but I have a few comments that I'll note.

smcabrera commented on lib/active_interaction/filter.rb:
```diff
@@ -51,6 +51,13 @@ def factory(slug)
         CLASSES.fetch(slug) { raise MissingFilterError, slug.inspect }
       end
 
+      # Returns the list of allowed options for this filter type.
+      #
+      # @return [Array<Symbol>]
+      def allowed_options
```
I didn't replicate this comment on all the subclass methods, but I could if we think that's valuable

smcabrera commented:
Trying to see if I can close up some open PRs before the new year 😅 

I made a few changes in response to feedback

1. I turned the constant level out arguments into a class method so that it could be overridden. HT @antulik for the feedback
2. I defined a method on the superclass and allowed the individual filter subclasses to define their own set. This allows filters to find only the options that are valid for them, making this validation more precise. HT @AaronLasseigne for the feedback
3. Updated the exception raised to be the ArgumentError that Ruby itself raises for invalid keyword arguments. HT @AaronLasseigne for the feedback.

Also added something to the changelog, @AaronLasseigne may have thoughts on what's the best way to summarize this change as I don't think I've committed to this project before

smcabrera commented:
@AaronLasseigne @antulik I think I've addressed your feedback, see [comment above](https://github.com/AaronLasseigne/active_interaction/pull/586#issuecomment-3702872862)
